File: src/eQuantic.UI.Primitives/Sheet/SheetDocument.cs
namespace eQuantic.UI.Primitives;

/// <summary>
/// The sheet's DATA: a sparse map of cell values plus per-row/per-column sizes. Values are strings
/// — what the user typed is the truth a v1 spreadsheet keeps (a formula engine, number formats and
/// cell styling are explicitly LATER; the fence is stated, not implied). Mutable, owned by the
/// controller; every mutation goes through it so undo can capture the inverse.
/// </summary>
public sealed class SheetDocument
{
    private readonly Dictionary<int, string> _cells = new();
    private readonly Dictionary<int, float> _rowHeights = new();
    private readonly Dictionary<int, float> _colWidths = new();

    public SheetDocument(int rows = 1000, int cols = 26)
    {
        Rows = Math.Max(1, rows);
        Cols = Math.Max(1, Math.Min(cols, CellRef.MaxCols));
    }

    /// <summary>The grid's LOGICAL extent — what scrolling and headers show. Cells outside it hold
    /// nothing; growing it is free (the map is sparse).</summary>
    public int Rows { get; internal set; }
    public int Cols { get; internal set; }

    public const float DefaultRowHeight = 28f;
    public const float DefaultColWidth = 96f;

    /// <summary>Cells that actually HOLD something — iteration for save, copy, and bounds.</summary>
    public IReadOnlyDictionary<int, string> Cells => _cells;

    public string GetCell(CellRef cell) =>
        _cells.TryGetValue(cell.Key, out var value) ? value : "";

    /// <summary>
    /// Writes a cell DIRECTLY — the LOADING path (an app filling the sheet from its source), which
    /// is deliberately not undoable. A user's edit goes through <see cref="SheetController.SetCell"/>,
    /// which records the inverse. An empty value REMOVES the entry — sparse means empty is absent.
    /// </summary>
    public void SetCell(CellRef cell, string value)
    {
        if (value.Length == 0) _cells.Remove(cell.Key);
        else _cells[cell.Key] = value;
    }

    public float RowHeight(int row) =>
        _rowHeights.TryGetValue(row, out var height) ? height : DefaultRowHeight;

    public float ColWidth(int col) =>
        _colWidths.TryGetValue(col, out var width) ? width : DefaultColWidth;

    internal void SetRowHeight(int row, float height)
    {
        if (Math.Abs(height - DefaultRowHeight) < 0.01f) _rowHeights.Remove(row);
        else _rowHeights[row] = Math.Max(12f, height);
    }

    internal void SetColWidth(int col, float width)
    {
        if (Math.Abs(width - DefaultColWidth) < 0.01f) _colWidths.Remove(col);
        else _colWidths[col] = Math.Max(24f, width);
    }

    public CellRef Clamp(CellRef cell) => new(
        Math.Clamp(cell.Row, 0, Rows - 1),
        Math.Clamp(cell.Col, 0, Cols - 1));

    /// <summary>
    /// Whether the cell holds a value — the building block of Excel's Ctrl+arrow "jump to the edge
    /// of the data block" motion.
    /// </summary>
    public bool HasValue(CellRef cell) => _cells.ContainsKey(cell.Key);

    /// <summary>
    /// Shifts every populated cell, row size and column size to make room for (or close over)
    /// inserted/deleted rows or columns. <paramref name="atRow"/>/<paramref name="atCol"/> pick the
    /// axis; positive <paramref name="delta"/> inserts, negative deletes. The REMOVED band's cells
    /// are returned so the edit can restore them on undo.
    /// </summary>
    internal List<SheetCellSnapshot> ShiftRows(int atRow, int delta)
    {
        var removed = new List<SheetCellSnapshot>();
        var moved = new List<SheetCellSnapshot>();
        foreach (var (key, value) in _cells)
        {
            var cell = CellRef.FromKey(key);
            if (cell.Row < atRow) continue;
            if (delta < 0 && cell.Row < atRow - delta)
            {
                removed.Add(new SheetCellSnapshot(cell, value));
                continue;
            }
            moved.Add(new SheetCellSnapshot(cell, value));
        }
        foreach (var snapshot in removed) _cells.Remove(snapshot.Cell.Key);
        foreach (var snapshot in moved) _cells.Remove(snapshot.Cell.Key);
        foreach (var snapshot in moved)
            _cells[new CellRef(snapshot.Cell.Row + delta, snapshot.Cell.Col).Key] = snapshot.Value;

        ShiftSizes(_rowHeights, atRow, delta);
        Rows = Math.Max(1, Rows + delta);
        return removed;
    }

    internal List<SheetCellSnapshot> ShiftCols(int atCol, int delta)
    {
        var removed = new List<SheetCellSnapshot>();
        var moved = new List<SheetCellSnapshot>();
        foreach (var (key, value) in _cells)
        {
            var cell = CellRef.FromKey(key);
            if (cell.Col < atCol) continue;
            if (delta < 0 && cell.Col < atCol - delta)
            {
                removed.Add(new SheetCellSnapshot(cell, value));
                continue;
            }
            moved.Add(new SheetCellSnapshot(cell, value));
        }
        foreach (var snapshot in removed) _cells.Remove(snapshot.Cell.Key);
        foreach (var snapshot in moved) _cells.Remove(snapshot.Cell.Key);
        foreach (var snapshot in moved)
            _cells[new CellRef(snapshot.Cell.Row, snapshot.Cell.Col + delta).Key] = snapshot.Value;

        ShiftSizes(_colWidths, atCol, delta);
        Cols = Math.Max(1, Math.Min(Cols + delta, CellRef.MaxCols));
        return removed;
    }

    private static void ShiftSizes(Dictionary<int, float> sizes, int at, int delta)
    {
        if (sizes.Count == 0) return;
        var entries = new List<KeyValuePair<int, float>>(sizes);
        foreach (var entry in entries)
        {
            if (entry.Key < at) continue;
            sizes.Remove(entry.Key);
        }
        foreach (var entry in entries)
        {
            if (entry.Key < at) continue;
            if (delta < 0 && entry.Key < at - delta) continue;   // the deleted band's sizes go with it
            sizes[entry.Key + delta] = entry.Value;
        }
    }
}

/// <summary>One cell's address and value, captured for undo.</summary>
public readonly record struct SheetCellSnapshot(CellRef Cell, string Value);

File: src/eQuantic.UI.Runtime/src/shared/__transpiled__/SheetDocument.ts
import { $eq, CellRef, SheetCellSnapshot } from "@equantic/runtime";

export class SheetDocument {
    constructor(rows: number = 1000, cols: number = 26, props?: any) {
        this._cells = {}; this._rowHeights = {}; this._colWidths = {}; this.rows = Math.max(1, rows);
        this.cols = Math.max(1, Math.min(cols, 16384)); if (props && typeof props === 'object') Object.assign(this, props);
    }

    _cells: Record<string, any>;
    _rowHeights: Record<string, any>;
    _colWidths: Record<string, any>;
    static defaultRowHeight: number = 28;
    static defaultColWidth: number = 96;
    rows: number = 0;
    cols: number = 0;

    get cells(): any {
        return this._cells;
    }

    getCell(cell: CellRef) {
        let value: any; return (Object.prototype.hasOwnProperty.call(this._cells, cell.key) ? ((value = this._cells[cell.key]), true) : false) ? value : '';
    }

    setCell(cell: CellRef, value: string) {
        if (value.length === 0) delete this._cells[cell.key]; else this._cells[cell.key] = value;
    }

    rowHeight(row: number) {
        let height: any; return (Object.prototype.hasOwnProperty.call(this._rowHeights, row) ? ((height = this._rowHeights[row]), true) : false) ? height : SheetDocument.defaultRowHeight;
    }

    colWidth(col: number) {
        let width: any; return (Object.prototype.hasOwnProperty.call(this._colWidths, col) ? ((width = this._colWidths[col]), true) : false) ? width : SheetDocument.defaultColWidth;
    }

    setRowHeight(row: number, height: number) {
        if (Math.abs(height - SheetDocument.defaultRowHeight) < 0.01) delete this._rowHeights[row]; else this._rowHeights[row] = Math.max(12, height);
    }

    setColWidth(col: number, width: number) {
        if (Math.abs(width - SheetDocument.defaultColWidth) < 0.01) delete this._colWidths[col]; else this._colWidths[col] = Math.max(24, width);
    }

    clamp(cell: CellRef) {
        return new CellRef(Math.min(Math.max(cell.row, 0), this.rows - 1), Math.min(Math.max(cell.col, 0), this.cols - 1));
    }

    hasValue(cell: CellRef) {
        return Object.prototype.hasOwnProperty.call(this._cells, cell.key);
    }

    shiftRows(atRow: number, delta: number) {
        let removed: SheetCellSnapshot[] = [];
        let moved: SheetCellSnapshot[] = [];
        for (const [key, value] of $eq.entries(this._cells, true)) {
            let cell = CellRef.fromKey(key);
            if (cell.row < atRow) continue;
            if (delta < 0 && cell.row < atRow - delta) {
                removed.push(new SheetCellSnapshot(cell, value));
                continue;
            }
            moved.push(new SheetCellSnapshot(cell, value));
        }
        for (const snapshot of removed) delete this._cells[snapshot.cell.key];
        for (const snapshot of moved) delete this._cells[snapshot.cell.key];
        for (const snapshot of moved) this._cells[new CellRef(snapshot.cell.row + delta, snapshot.cell.col).key] = snapshot.value;
        SheetDocument.shiftSizes(this._rowHeights, atRow, delta);
        this.rows = Math.max(1, this.rows + delta);
        return removed;
    }

    shiftCols(atCol: number, delta: number) {
        let removed: SheetCellSnapshot[] = [];
        let moved: SheetCellSnapshot[] = [];
        for (const [key, value] of $eq.entries(this._cells, true)) {
            let cell = CellRef.fromKey(key);
            if (cell.col < atCol) continue;
            if (delta < 0 && cell.col < atCol - delta) {
                removed.push(new SheetCellSnapshot(cell, value));
                continue;
            }
            moved.push(new SheetCellSnapshot(cell, value));
        }
        for (const snapshot of removed) delete this._cells[snapshot.cell.key];
        for (const snapshot of moved) delete this._cells[snapshot.cell.key];
        for (const snapshot of moved) this._cells[new CellRef(snapshot.cell.row, snapshot.cell.col + delta).key] = snapshot.value;
        SheetDocument.shiftSizes(this._colWidths, atCol, delta);
        this.cols = Math.max(1, Math.min(this.cols + delta, 16384));
        return removed;
    }

    static shiftSizes(sizes: Record<string, any>, at: number, delta: number) {
        if (Object.keys(sizes).length === 0) return;
        let entries = $eq.entries(sizes, true);
        for (const entry of entries) {
            if (entry.key < at) continue;
            delete sizes[entry.key];
        }
        for (const entry of entries) {
            if (entry.key < at) continue;
            if (delta < 0 && entry.key < at - delta) continue;
            sizes[entry.key + delta] = entry.value;
        }
    }
}

